scripts/rust-libc.py

```diff
@@ -0,0 +1,581 @@
+					typename = "u16"
+				elif self.type.get_declaration().displayname in ("int16_t", "__mlibc_int16"):
+					typename = "i16"
+				elif self.type.get_declaration().displayname in ("uint32_t", "__mlibc_uint32"):
+					typename = "u32"
+				elif self.type.get_declaration().displayname in ("int32_t", "__mlibc_int32"):
+					typename = "i32"
+				elif self.type.get_declaration().displayname in ("uint64_t", "__mlibc_uint64"):
+					typename = "u64"
+				elif self.type.get_declaration().displayname in ("int64_t", "__mlibc_int64"):
+					typename = "i64"
+				elif self.type.get_declaration().displayname in ("intptr_t", "__mlibc_intptr"):
+					typename = "isize"
+				elif self.type.get_declaration().displayname in ("__mlibc_size"):
+					typename = "usize"
+				else:
+					typename = "::" + str(self.type.get_declaration().displayname)
+			case TypeKind.POINTER:
+				typename = self.convert_ptr_type(self.cursor, self.type)
+			case TypeKind.TYPEDEF:
+				return str(self.type.spelling)
+			case TypeKind.RECORD:
+				return ""
+		return typename
+
+	@property
+	def canonical(self):
+		return str(Type(self.type.get_canonical()))
+
+	def is_valid(self):
+		return self.kind != TypeKind.INVALID
+
+	def is_va_list(self):
+		return self.kind == TypeKind.ELABORATED and self.type.get_declaration().displayname == "__builtin_va_list"
+
+	def escape_name(name: str):
+		if name in ("type", "in"):
+			return f"r#{name}"
+		return name
+
+	def cursor_name(c: Cursor):
+		d = c.type.get_declaration()
+		if d and d.is_anonymous():
+			return f"anon_{pathlib.Path(str(d.location.file)).stem}_line{d.location.line}"
+		return Type.escape_name(c.displayname)
+
+class State:
+	def __init__(self):
+		self.functions = []
+		self.macros = []
+		self.types = []
+		self.structs = []
+		self.variables = []
+
+class RustBindingGenerator:
+	def __init__(self, config):
+		self.config = config
+		self.in_function_block = False
+		self.in_struct_block = False
+		self.in_union_block = False
+
+	def handle_macro(self, cursor, gen, state):
+		def is_num(s):
+			if s.removeprefix("0o").isnumeric():
+				return True
+			if set(s.removeprefix("0x")).issubset(string.hexdigits):
+				return True
+			return False
+
+		done = False
+		is_negative = False
+
+		assert(len(gen) >= 1)
+		assert(gen[0].kind == TokenKind.IDENTIFIER)
+		gen.pop(0)
+
+		if len(gen) >= 1:
+			tokens = []
+			c_type = "::c_int"
+			is_unsigned = False
+			i = 0
+			while not done and gen and i < len(gen):
+				c_type = "int"
+				if gen[i].kind == TokenKind.PUNCTUATION and gen[i].spelling in ("(", ")"):
+					if not (i == 0 or i == (len(gen) - 1)):
+						tokens.append(gen[i].spelling)
+					i += 1
+				elif gen[i].kind == TokenKind.PUNCTUATION and gen[i].spelling == "-":
+					is_unsigned = False
+					i += 1
+				elif gen[i].kind in (TokenKind.LITERAL, TokenKind.IDENTIFIER, TokenKind.PUNCTUATION):
+					spelling = gen[i].spelling
+					if spelling.endswith("ULL") and is_num(spelling[:-3]):
```

Comment:
this can definitely be simpler

Reply:
I don't see how, as this is simple and quite clear